### ISSUE TITLE
How to discriminate a polymorphic class with enum?

### ISSUE THREAD

**Issue by THeK3nger** [cbeust/klaxon#287]
This is more a question than an issue. Not sure if this is possible in the current Klaxon version.

I am trying to deserialize a polymorphic class using an enum and not a string. 

This is what I want.

```kotlin
@TypeFor(field = "type", adapter = ResponseTypeAdapter::class)
open class Response(val type: ResponseType, platform: MessagingPlatform? = null)

data class TextResponse(
    val platform: MessagingPlatform,
    val options: List<String>
) : Response(ResponseType.TEXT, platform)

data class CarouselResponse(
    val platform: MessagingPlatform,
    val cardsArray: List<Any>
) : Response(ResponseType.CAROUSEL, platform)
```

where `ResponseType` is an enum constant like

```
enum class ResponseType {
    TEXT,
    VIDEO,
    QUICKANSWERS,
    CAROUSEL,
    IMAGE,
    SHARE,
    HANDOVER,
    // ...
}
```

Of course, I cannot create a `TypeAdapter` like the one below because I think Klaxon will still use the original string in the `Response` initialization. (In fact, I get the error: `java.lang.IllegalArgumentException: Could not find enum value for text`. Note that `text` is lower case like in the JSON).

```
class ResponseTypeAdapter: TypeAdapter<Response> {
    override fun classFor(type: Any): KClass<out Response> {
        val responseType = ResponseType.valueOf((type as String).toUpperCase())
        return when(responseType) {
            ResponseType.TEXT -> TextResponse::class
            ResponseType.CAROUSEL -> CarouselResponse::class
           // all the other cases
        }
    }
}
```

This works only if the enum and the string in the JSON are represented by the same string constant. For instance, if my enum is defined in lower case like the string in the JSON, the deserialization works fine.

```kotlin
enum class ResponseType {
    text,
    image,
    carousel,
    // ... 
} // This enum works.
```

So, in conclusion, I am asking if there is a way to discriminate using the `TypeAdapter` but returning also a converted discriminant `type`.

**Comment by cbeust** [comment]
I think it's dangerous to rely on the spelling of your `enum` values since a harmless refactoring will break your JSON parsing.

How about just implementing your adapter so it transforms your enum values in strings that you control and which don't rely on what the enum values actually are?

**Comment by THeK3nger** [comment]
> How about just implementing your adapter so it transforms your enum values in strings that you control and which don't rely on what the enum values actually are?

That's where I have trouble, probably because I am missing something on how to use effectively `TypeAdapter`. Why should I convert enum into strings and not string into enum?

Let's see if I can explain my line of thoughts. In the `when` clause I think that each line is `"string_value_in_the_json" ->  ClassToUse::class`. Then, Klaxon will instantiate a `ClassToUse::class` accordingly.

So, this is what I have now as a workaround. This is the type adapter

```kotlin
class ResponseTypeAdapter : TypeAdapter<Response> {
    override fun classFor(type: Any): KClass<out Response> {
        return when (type as String) {
            "text" -> TextResponse::class
            "video" -> VideoResponse::class
            "quickanswers" -> QuickAnswerResponse::class
            "carousel" -> CarouselResponse::class
            "image" -> ImageResponse::class
            "share" -> ShareResponse::class
            else -> TODO()
        }
    }
}
```

But then I needed to modify the constructor or each `Response` into exposing a constructor with a `type: String` that internally convert the string into the appropriate enum value. However, I feel is weird to expose a constructor just for the json parser. It would be better if I can convert tell to the type adapter that, for instance: "if type has a string value of "text" then instantiate a `TextResponse` **and convert type using this `StringTypeToEnum` function**. Where `StringTypeToEnum` is a function I can write and control that decouple the enum from the spelling of the string value. 

Now that I am writing this, I wonder if using field conversion overriding may be a better way to do it.

Sorry if this does not make any sense. I may be approaching this problem from a totally wrong angle. :)
